lib/quic.c

```diff
@@ -0,0 +1,983 @@
+  (void)stream_user_data;
+
+  /* TODO: implement */
+
+  return 0;
+}
+
+static int cb_stream_close(ngtcp2_conn *tconn, uint64_t stream_id,
+                           uint16_t app_error_code,
+                           void *user_data, void *stream_user_data)
+{
+  struct connectdata *conn = (struct connectdata *)user_data;
+  (void)conn;
+  (void)tconn;
+  (void)stream_id;
+  (void)app_error_code;
+  (void)stream_user_data;
+  /* stream is closed... */
+
+  return 0;
+}
+
+static int cb_recv_retry(ngtcp2_conn *tconn, const ngtcp2_pkt_hd *hd,
+                         const ngtcp2_pkt_retry *retry, void *user_data)
+{
+  /* Re-generate handshake secrets here because connection ID might change. */
+  struct connectdata *conn = (struct connectdata *)user_data;
+  (void)tconn;
+  (void)hd;
+  (void)retry;
+
+  quic_init_ssl(conn);
+  setup_initial_crypto_context(conn);
+
+  return 0;
+}
+
+static int cb_extend_max_stream_id(ngtcp2_conn *tconn, uint64_t max_stream_id,
+                                   void *user_data)
+{
+  struct connectdata *conn = (struct connectdata *)user_data;
+  /* not clear what this wants */
+  (void)max_stream_id;
+  (void)tconn;
+  (void)conn;
+
+  return 0;
+}
+
+
+static void quic_callbacks(ngtcp2_conn_callbacks *c)
+{
+  memset(c, 0, sizeof(ngtcp2_conn_callbacks));
+  c->client_initial = cb_initial;
+  c->recv_crypto_data = cb_recv_crypto_data;
+  c->handshake_completed = cb_handshake_completed;
+  /* recv_version_negotiation = NULL */
+  c->in_encrypt = cb_in_encrypt;
+  c->in_decrypt = cb_in_decrypt;
+  c->encrypt = cb_encrypt_data;
+  c->decrypt = cb_decrypt_data;
+  c->in_encrypt_pn = cb_in_encrypt_pn;
+  c->encrypt_pn = cb_encrypt_pn;
+  c->recv_stream_data = cb_recv_stream_data;
+  c->acked_crypto_offset = cb_acked_crypto_offset;
+  c->acked_stream_data_offset = cb_acked_stream_data_offset;
+  c->stream_close = cb_stream_close;
+  c->recv_retry = cb_recv_retry;
+  c->extend_max_stream_id = cb_extend_max_stream_id;
+}
+
+
+CURLcode Curl_quic_connect(struct connectdata *conn,
+                           curl_socket_t sockfd,
+                           const struct sockaddr *addr,
+                           socklen_t addrlen)
+{
+  int rc;
+  struct quicsocket *qs = &conn->quic;
+  (void)sockfd;
+  (void)addr;
+  (void)addrlen;
+  infof(conn->data, "Connecting socket %d over QUIC\n", sockfd);
+
+  qs->sslctx = quic_ssl_ctx(conn->data);
+  if(!qs->sslctx)
+    return CURLE_FAILED_INIT; /* TODO: better return code */
+
+  if(quic_init_ssl(conn))
+    return CURLE_FAILED_INIT; /* TODO: better return code */
+
+  quic_settings(&qs->settings);
+  quic_callbacks(&qs->callbacks);
+
+#ifdef NGTCP2_PROTO_VER_D15
+#define QUICVER NGTCP2_PROTO_VER_D15
+#else
+#define QUICVER NGTCP2_PROTO_VER_D14
+#endif
+  /* ngtcp2 master branch uses version NGTCP2_PROTO_VER_D14 */
+  rc = ngtcp2_conn_client_new(&qs->conn, &qs->dcid, &qs->scid,
+                              QUICVER, &qs->callbacks, &qs->settings, conn);
```

Comment:
It looks like qs->dcid and qs->scid are not initialized.
They must be some random bytes.